rbpf-cli/src/coverage.rs

```diff
@@ -0,0 +1,257 @@
+use {
+    crate::gcov::{GcovFile, GcovIntermediate, GcovLine},
+    bv::{BitVec, BitsMut},
+    gimli::{Dwarf, EndianSlice, LineProgramHeader, RunTimeEndian, Unit, UnitHeader},
+    goblin::elf::Elf,
+    itertools::Itertools,
+    log::*,
+    std::{
+        borrow::Cow,
+        collections::{BTreeSet, HashMap},
+        fmt::{Debug, Formatter},
+        path::{Path, PathBuf},
+    },
+};
+
+#[derive(Default)]
+pub(crate) struct FileCoverage {
+    file_path: Option<PathBuf>,
+    hits: BTreeSet<(u64, u64)>,
+}
+
+impl FileCoverage {
+    pub(crate) fn new(file_path: Option<PathBuf>) -> Self {
+        Self {
+            file_path,
+            ..FileCoverage::default()
+        }
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct Coverage {
+    hits: HashMap<u64, FileCoverage>,
+}
+
+impl Coverage {
+    pub(crate) fn from_trace(
+        elf_bytes: &[u8],
+        elf: &Elf<'_>,
+        trace: &[[u64; 12]],
+    ) -> Result<Self, Box<dyn std::error::Error>> {
+        // Find text section.
+        let text_range = elf
+            .section_headers
+            .iter()
+            .find(|section| elf.shdr_strtab.get_at(section.sh_name) == Some(".text"))
+            .ok_or("missing .text section")?
+            .file_range()
+            .ok_or("invalid .text range")?;
+
+        // Create bitmap of executed instructions.
+        let mut hits = BitVec::<usize>::new_fill(false, (text_range.len() / 8) as u64);
+        for ins in trace {
+            hits.set_bit(ins[11], true);
```

Comment:
Using a bitmap is not ideal because it does not accurately count the number of times a line was executed. It would be better to use `Vec<usize>` here. I aim to fix this in a follow-up PR unless this is a blocker for merging.